### DISCUSSION TITLE
Allow usage of RKE2, instead of K3s

### DISCUSSION THREAD

**Discussion by beisenherz** [mysticaltech/terraform-hcloud-kube-hetzner#1132] [Ideas]
### Description

K3s comes with some patches, that cause the release cycle of K3s to be slower than RKE2. Also does RKE2 come with an FIPS 140-2 compliance, and a compatible CLI to the one of K3s. I believe this change would be relatively fast to implement.
Would this be of any interest for this project?

**Comment by mysticaltech** [comment]
@Sironheart  Most definitely this is something we would like to support. We actually thought about it in the past, but it requires a decent amount of work to integrating and have a switching logic between the two. If it's something you want to give a shot, me and the @kube-hetzner/core  would be more than happy to assist you. PRs welcome (we can work on a draft PR together and iterate our way to supporting RKE2)!

**Reply by valkenburg-prevue-ch** [reply]
One reason which kept me from this feature, in that past, was that at that time the SELinux patches for RKE2 were not kept up to date as quickly as those for k3s. Perhaps that has changed now?

**Comment by codeagencybe** [comment]
Did somebody already start working on this? Because I'm also much interested in RKE2 over k3s. 
RKE2 is way more conformant than k3s and is a lot more closer/uptodate to native Kubernetes than k3s. 

I even had some people telling me for security purpose, it should be RK2E by default. 
K3s has it's benefits for airgapped installs and being a tiny bit smaller than RKE2 that is beneficial to run on low power/low resource infra like Raspberry Pi's etc... 

I have checked some differences about the installation process between k3s and RKE2 and so far I think @Sironheart is right about being a minimal impact to changes. It's mostly changing endpoints where to fetch the downloaders.
And one big change is that RKE2 does not autostart after installation and requires a manual "kick off". 

If there is already a start somewhere, please share as I want to help to make this happen.

**Reply by mysticaltech** [reply]
@codeagencybe Indeed, you are correct. We are currently working on v3 with @aleksasiriski, see #1164,  that will make the project way more modular, and will allow an easy optional swap of k3s to rke2 as a later feature. So first thing for us is getting v3 out, after which we will focus on supporting rke2 and help would be more than appreciated.

**Reply by codeagencybe** [reply]
OK, that's nice news! Any rough estimation on timeframe when v3 is expected as "ready"? 
Anything I can help with? 

Anybody here who has used Atlantis before? Open Source Terraform Pull Request Automation
https://www.runatlantis.io/

**Reply by aleksasiriski** [reply]
I'm working on a project for client currently, after that I will focus on v3 refactor. We don't have an estimate.

As for atlantis, looks cool, but is most likely not suitable for this project since it's very expensive to run a cluster that does nothing but run PR terraform plans.. also, it wouldn't be possible to have one cluster for all the options available in this project.

**Reply by mikeywuu** [reply]
Is there any progress made here? I would also be willing to help to adapt to RKE2. I mainly use rancher managed downstream clusters, also with RKE2, in "normal" on-premise VMs and would like to switch to this project, avoiding re-writing basically many features.

**Reply by aleksasiriski** [reply]
I'm currently rewriting everything from scratch, with focus on k3s but it should be easy to add RKE2 later as an option.

**Reply by mysticaltech** [reply]
@mikeywuu No progress yet, but if you want to give it a shot based on the present codebase, while @aleksasiriski works on the new architecture, please be our guest! 

Since k3s and rke2 both follow the same config format and logic, having as DRY as possible code between the two with a proper switching logic should be very doable.

**Reply by mikeywuu** [reply]
@mysticaltech I would love to give it a try. Are there any special things I should consider regarding the ongoing refactoring in the background?

Also, I was wondering if the refactoring will give me the possibility to create a "management" rancher cluster which creates several "downstream" clusters. Currently, I can pass the URLs for joining / bootstrapping the management cluster for rancher to my own RKE2 installation logic.

**Reply by aleksasiriski** [reply]
I've not yet dabbled in Rancher, since I don't use it personally. The refactor should give almost zero problems for using RKE2 instead of k3s, but as for Rancher I don't really know. Anyways, since everything is modularized it will allow to easily inject whatever else is needed to make Rancher work so it shouldn't be much of a problem to add it later on :) I will be open sourcing the refactored code at the start of September.

**Reply by mikeywuu** [reply]
Hey @aleksasiriski, is there an estimate when we can expect some of the modularization? Will this be compatible with already existing clusters deployed via this project or is a re-deployment necessary? I would assume the latter?

**Reply by aleksasiriski** [reply]
No estimate yet to publish, but it's progressing very good. It will most definately have to be redeployment since the architecture is completely different and a lot of the features are missing (will have to be deployed manually, like Longhorn etc.)

**Reply by maggie44** [reply]
@aleksasiriski is the a branch you are working on for the refactor? Would be interested in taking a look at what a RKE2 implementation would involve, but seems not much point doing it against main branch right now if there is going to be a big refactor coming.

**Reply by tiran133** [reply]
Is there any update on this?

**Reply by mikeywuu** [reply]
@aleksasiriski can we help? Still excited to see this big chance coming :)

**Comment by mikeywuu** [comment]
Hi folks,
im actively working on this feature and have made already good progress on it. Im able to spawn a RKE2 cluster with cilium. Existing k3s clusters are unaffected as far as I could test.
Im close to opening a draft PR along some topics open for discussion. I tried / am trying to keep it as DRY as possible but unfortunately some things are bootstrapped differently with RKE2.

I would like to open this up for others to contribute to make sure other scenarios can be tested / fixed. Should I open a draft PR for this so we can work on this together?

**Reply by mysticaltech** [reply]
@mikeywuu I want to support it, please shoot your PR as is, we can then clean it up with things like openai codex.

**Reply by mikeywuu** [reply]
@mysticaltech sure  - let me rebase my branch and make sure that my changes still work as expected, then I will go ahead and open the PR. I have a list of things that are still open or not yet implemented, following a couple of questions. Is there a way we can get into contact so we can also work on the PR together? I am currently already running a cluster with workloads on it, so at least for the changes I'm confident that they are stable and work with upgrades/restores etc.

**Reply by mysticaltech** [reply]
@mikeywuu Appreciate it. Emails would have been natural but I don't check them too often. My prefered way is via a draft PR, and we can work on it async. I have a few ideas how how I would like it to work, so if you modified this project to add it, that's all I need. I need a working version and the rest can be retraced back. I use AI a lot.

**Reply by mikeywuu** [reply]
@mysticaltech / everyone interested, the PR #1784 is now created, Im happy for any feedback and help to finish this :)

**Reply by mysticaltech** [reply]
Awesome @mikeywuu , thanks for sharing 🙏 I will have a good look at the PR ASAP, seems like amazing work already
